uPortal-content/uPortal-content-portlet/src/main/java/org/apereo/portal/portlet/dao/jpa/PortletDefinitionImpl.java

```diff
@@ -714,18 +665,59 @@ public EntityIdentifier getEntityIdentifier() {
 
     @Override
     public PortletLifecycleState getLifecycleState() {
+
+        /*
+         * NB:  A portlet definition may have a current lifecycle
+         * state of null if it has not yet been completely created.
+         */
+        PortletLifecycleState rslt = null;
+
         final Date now = new Date();
-        if (parameters.containsKey(PortletLifecycleState.MAINTENANCE_MODE_PARAMETER_NAME)) {
-            return PortletLifecycleState.MAINTENANCE;
-        } else if (expirationDate != null && expirationDate.before(now)) {
-            return PortletLifecycleState.EXPIRED;
-        } else if (publishDate != null && publishDate.before(now)) {
-            return PortletLifecycleState.PUBLISHED;
-        } else if (approvalDate != null && approvalDate.before(now)) {
-            return PortletLifecycleState.APPROVED;
-        } else {
-            return PortletLifecycleState.CREATED;
+        for (IPortletLifecycleEntry entry : lifecycleEntries) {
+            /*
+             * A lifecycle change that is scheduled for the
+             * future is not a candidate for the current state.
+             */
+            if (entry.getDate().before(now)) {
+                rslt = entry.getLifecycleState();
+            } else {
+                /*
+                 * NOTE (per JPA docs):  "If the ordering element is not specified,
+                 * ordering by the primary key of the associated entity is assumed."
+                 */
+                break;
+            }
         }
+
+        return rslt;
+
+    }
+
+    @Override
+    public void setLifecycleState(PortletLifecycleState lifecycleState, IPerson user) {
+        setLifecycleState(lifecycleState, user, new Date());
+    }
+
+    @Override
+    public void setLifecycleState(PortletLifecycleState lifecycleState, IPerson user, Date timestamp) {
```

Comment:
Intended concurrency semantics? What's the intention around e.g. multiple threads trying to set lifecycle state concurrently?